### PR TITLE
ENH: linalg: Cythonize `id_dist` FORTRAN code

### LINKED ISSUE #4275: scipy.linalg.interpolative svd  integer overflow
I am working with very large sparse matrix (50k x 1.5M) and trying to obtain rank 500 svd.
In the signature for iddr_rsvd.f in scipy/scipy/linalg/interpolative.pyf the intermediate calculation of size for `w` matrix overflows

```
 w((krank+1)*(2*m+4*n)+25*pow(krank,2))
```

https://github.com/scipy/scipy/blob/master/scipy/linalg/interpolative.pyf#L325.

I just built scipy 0.14 by changing the faulty line to 

```
 w((krank+1)*(2ULL*m+4ULL*n)+25*pow(krank,2))
```

and confirmed that it does not overflow.  I am not sure if this is the best way to fix this problem as I have very little experience with f2py. Another possibility is to change the integer types for variable `m` and `n`.

### LINKED ISSUE #9249: scipy.linalg.tests.test_interpolative::TestInterpolativeDecomposition::test_id fails on 32-bit intel
### Reproducing code example:
```
python3.6 runtests.py -v -t scipy.linalg.tests.test_interpolative::TestInterpolativeDecomposition::test_id
```

### Error message:

```
Building, see build.log...
Build OK (0:00:02.503862 elapsed)
============================= test session starts ==============================
platform linux -- Python 3.6.6+, pytest-3.6.4, py-1.5.4, pluggy-0.6.0 -- /usr/bin/python3.6
cachedir: ../../../../../.pytest_cache
rootdir: /root/scipy, inifile: pytest.ini
collecting ... collected 1 item

scipy/linalg/tests/test_interpolative.py::TestInterpolativeDecomposition::test_id FAILED [100%]

=================================== FAILURES ===================================
____________________ TestInterpolativeDecomposition.test_id ____________________

self = <scipy.linalg.tests.test_interpolative.TestInterpolativeDecomposition object at 0xf4ee6cac>

    def test_id(self):
        for dtype in [np.float64, np.complex128]:
>           self.check_id(dtype)

dtype      = <class 'numpy.complex128'>
self       = <scipy.linalg.tests.test_interpolative.TestInterpolativeDecomposition object at 0xf4ee6cac>

scipy/linalg/tests/test_interpolative.py:47: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <scipy.linalg.tests.test_interpolative.TestInterpolativeDecomposition object at 0xf4ee6cac>
dtype = <class 'numpy.complex128'>

    def check_id(self, dtype):
        # Test ID routines on a Hilbert matrix.
    
        # set parameters
        n = 300
        eps = 1e-12
    
        # construct Hilbert matrix
        A = hilbert(n).astype(dtype)
        if np.issubdtype(dtype, np.complexfloating):
            A = A * (1 + 1j)
        L = aslinearoperator(A)
    
        # find rank
        S = np.linalg.svd(A, compute_uv=False)
        try:
            rank = np.nonzero(S < eps)[0][0]
        except IndexError:
            rank = n
    
        # print input summary
        _debug_print("Hilbert matrix dimension:        %8i" % n)
        _debug_print("Working precision:               %8.2e" % eps)
        _debug_print("Rank to working precision:       %8i" % rank)
    
        # set print format
        fmt = "%8.2e (s) / %5s"
    
        # test real ID routines
        _debug_print("-----------------------------------------")
        _debug_print("Real ID routines")
        _debug_print("-----------------------------------------")
    
        # fixed precision
        _debug_print("Calling iddp_id / idzp_id  ...",)
        t0 = time.time()
        k, idx, proj = pymatrixid.interp_decomp(A, eps, rand=False)
        t = time.time() - t0
        B = pymatrixid.reconstruct_matrix_from_id(A[:, idx[:k]], idx, proj)
        _debug_print(fmt % (t, np.allclose(A, B, eps)))
>       assert_(np.allclose(A, B, eps))
E       AssertionError

A          = array([[1.        +1.j        , 0.5       +0.5j       ,
        0.33333333+0.33333333j, ..., 0.0033557 +0.0033557j ,
 ...
        0.00331126+0.00331126j, ..., 0.00167504+0.00167504j,
        0.00167224+0.00167224j, 0.00166945+0.00166945j]])
B          = array([[1.        +1.00000000e+00j, 0.5       +5.00000000e-01j,
        0.33333333+3.33333333e-01j, ..., 0.00252299+3....126+3.31125828e-03j, ..., 0.00240064+4.01219469e-04j,
        0.00239738+3.93272780e-04j, 0.00239412+3.85406210e-04j]])
L          = <300x300 MatrixLinearOperator with dtype=complex128>
S          = array([3.28383209e+00, 1.46118973e+00, 4.85550807e-01, 1.40695788e-01,
       3.77404295e-02, 9.57267036e-03, 2.318987...3.61515144e-19, 3.44841007e-19, 2.67885523e-19,
       2.14393434e-19, 1.38418788e-19, 1.08774972e-19, 3.53735496e-20])
dtype      = <class 'numpy.complex128'>
eps        = 1e-12
fmt        = '%8.2e (s) / %5s'
idx        = array([  0,  76,   1, 134,   2,   5,   6,   7,   8,   9,  10,  11,  12,
        13,  14,  15,  16,  17,  18,  19,  20,..., 282, 283, 284, 285,
       286, 287, 288, 289, 290, 291, 292, 293, 294, 295, 296, 297, 298,
       299], dtype=int32)
k          = 5
n          = 300
proj       = array([[ 1.14253932e+00+8.14415919e-01j,  1.02685117e+00+7.27996082e-01j,
         9.31616510e-01+6.56616558e-01j, .......,
        -6.28869125e-03-4.88365418e-03j, -6.30277694e-03-4.85859879e-03j,
        -6.31662104e-03-4.83374460e-03j]])
rank       = 20
self       = <scipy.linalg.tests.test_interpolative.TestInterpolativeDecomposition object at 0xf4ee6cac>
t          = 0.0024597644805908203
t0         = 1536620034.2090552

scipy/linalg/tests/test_interpolative.py:89: AssertionError
=========================== 1 failed in 0.55 seconds ===========================
```

### Scipy/Numpy/Python version information:
```
1.2.0.dev0+641bc92 1.14.5 sys.version_info(major=3, minor=6, micro=6, releaselevel='final', serial=0)
```

### LINKED ISSUE #12593: scipy.linalg.interpolative.estimate_rank is quite unreliable
Even though the [documentation](https://scipy.github.io/devdocs/generated/scipy.linalg.interpolative.estimate_rank.html#scipy.linalg.interpolative.estimate_rank) of ```scipy.linalg.interpolative.estimate_rank``` states "If A is of type numpy.ndarray, then the output rank is typically about 8 higher than the actual numerical rank.", the following code snippet highlights its unreliability:

```python
import numpy as np
from scipy.linalg import interpolative

def ID_ranks(shape, runs, eps):
    np.random.seed(0)
    for i in range(runs):
        A = np.random.rand(shape[0], shape[1])
        Ap = A @ A.T
        estimated_rank = interpolative.estimate_rank(Ap, eps)
        print("True rank:", min(shape), "Estimated rank from Interpolative Decomposition:", estimated_rank)

ID_ranks([100,52], 10, 1e-13)
ID_ranks([1000,500], 10, 1e-13)
```
For the 100x52 case, ID's output is always 100 and this is the case for every row value >= 52. For the 1000x500 case, ID's output is always 1000 and this is the case for every row value >= 500. No matter how small ```eps``` I set, the results are always the same.

### LINKED ISSUE #18367: RFC: Get rid of linalg.interpolative Fortran code
I have been working on the linalg/src directory for a while now. The `id` library is written in Fortran and in a de-facto unmaintained status. While I am grateful for the authors' efforts back in the day, we were a bit too liberal in including Fortran code and now we found ourselves in a strange predicament. 

The code doesn't do much but deals with LinearOperator and sparse arrays. The `id` code is very unorthodox by today's standards and I don't have too much of hope to modernize it through Fortran tweaks. Plus it's randomization machinery is soon going to be noncompliant with NumPy (I just learned about these things very recently so allow me to pretend knowledge here). 

I can go about my Fortran diatribe for a long thread but TL;DR We have to get Fortran parts out of the way and `f2py` is not cutting it anymore.

To give a single example if you want to support ndarrays, F-arrays are useless because you can't slice them in the last dimensions. That is to say `(x, y, :, :)` is not contiguous but  `(:, :, x, y)` is for F-ndarrays. OK but then we want to use some obscure Fortran library but still keep the C dominated NumPy operations what gives? Welcome to the world of buffer joggling. In fact just because of the wrappers we tell the users [to start with an F array](https://docs.scipy.org/doc/scipy/reference/linalg.interpolative.html#initializing). 

I am not writing this to be controversial but we have to get rid of all Fortran libraries and while we are at it all `numpy.matrix` usage at some point. It is not an impossible task but we have to warm up to this idea. After fighting with f2py and all fortran tooling since the lockdown, I can't see any benefit any more other than continuous headache and causing noncompliancy across the board. `f2py` is seriously dear to my heart however it is showing its age and there is no light at the tunnel yet to modernize it and address many performance issues that come with it. 

Let me emphasize, Fortran is a nice and a very fast language. But that is all there is to say about it. Everything else from our side is just pain to work with it. And hence just to kick start this crusade, I would like to propose that we get rid of this codebase. In fact maybe this code should be moved to `scipy.sparse.linalg` completely.

### LINKED ISSUE #20452: BUG:linalg:interpolative: Crashing if k is requested too high in `interp_decomp`
### Describe your issue.

I am rewriting `linalg.interpolative` Fortran code for #18367 and started discovering some edges. I don't know what the cause of this but probably some wrapper allocation issue. Here is just to record in time in case later we stumble upon it though I highly doubt anyone is using this. I'll close this with the rewrite PR. 

### Reproducing Code Example

```python
import scipy.linalg.interpolative as lai
import numpy as np

A = np.random.rand(8, 6)
lai.interp_decomp(A, eps_or_k=10)
```

### Error message

```shell
The kernel died, restarting...
```

### SciPy/NumPy/Python version and system information

```shell
1.13.0 1.26.4 sys.version_info(major=3, minor=12, micro=1, releaselevel='final', serial=0)
Build Dependencies:
  blas:
    detection method: pkgconfig
    found: true
    include directory: /c/opt/64/include
    lib directory: /c/opt/64/lib
    name: openblas
    openblas configuration: USE_64BITINT= DYNAMIC_ARCH=1 DYNAMIC_OLDER= NO_CBLAS=
      NO_LAPACK= NO_LAPACKE= NO_AFFINITY=1 USE_OPENMP= ZEN MAX_THREADS=24
    pc file directory: c:/opt/64/lib/pkgconfig
    version: 0.3.26.dev
  lapack:
    detection method: pkgconfig
    found: true
    include directory: /c/opt/64/include
    lib directory: /c/opt/64/lib
    name: openblas
    openblas configuration: USE_64BITINT= DYNAMIC_ARCH=1 DYNAMIC_OLDER= NO_CBLAS=
      NO_LAPACK= NO_LAPACKE= NO_AFFINITY=1 USE_OPENMP= ZEN MAX_THREADS=24
    pc file directory: c:/opt/64/lib/pkgconfig
    version: 0.3.26.dev
  pybind11:
    detection method: config-tool
    include directory: unknown
    name: pybind11
    version: 2.12.0
Compilers:
  c:
    commands: cc
    linker: ld.bfd
    name: gcc
    version: 10.3.0
  c++:
    commands: c++
    linker: ld.bfd
    name: gcc
    version: 10.3.0
  cython:
    commands: cython
    linker: cython
    name: cython
    version: 3.0.10
  fortran:
    commands: gfortran
    linker: ld.bfd
    name: gcc
    version: 10.3.0
  pythran:
    include directory: C:\Users\runneradmin\AppData\Local\Temp\pip-build-env-tun37eh0\overlay\Lib\site-packages/pythran
    version: 0.15.0
Machine Information:
  build:
    cpu: x86_64
    endian: little
    family: x86_64
    system: windows
  cross-compiled: false
  host:
    cpu: x86_64
    endian: little
    family: x86_64
    system: windows
Python Information:
  path: C:\Users\runneradmin\AppData\Local\Temp\cibw-run-on03_hml\cp312-win_amd64\build\venv\Scripts\python.exe
  version: '3.12'
```

### LINKED ISSUE #20609: BUG:linalg:interpolative: Inputs are not mapped to compatible dtypes
### Describe your issue.

Integer inputs trip up main routine.

### Reproducing Code Example

```python
import numpy as np
import scipy.linalg.interpolative as lai

zzz = np.arange(36).reshape(6, 6)
lai.interp_decomp(zzz, eps_or_k=1e-5, rand=False)
# or alternatively
lai.interp_decomp(zzz, eps_or_k=1e-5, rand=True)
# or alternatively
lai.interp_decomp(zzz, eps_or_k=3, rand=False)
```

### Error message

```shell
ValueError: invalid input dtype (input must be float64 or complex128)
```

### SciPy/NumPy/Python version and system information

<details>

<summary>Sys Info</summary>

```shell
1.12.0 1.26.4 sys.version_info(major=3, minor=10, micro=12, releaselevel='final', serial=0)
Build Dependencies:
  blas:
    detection method: pkgconfig
    found: true
    include directory: /usr/local/include
    lib directory: /usr/local/lib
    name: openblas
    openblas configuration: USE_64BITINT= DYNAMIC_ARCH=1 DYNAMIC_OLDER= NO_CBLAS=
      NO_LAPACK= NO_LAPACKE= NO_AFFINITY=1 USE_OPENMP= HASWELL MAX_THREADS=2
    pc file directory: /usr/local/lib/pkgconfig
    version: 0.3.21.dev
  lapack:
    detection method: pkgconfig
    found: true
    include directory: /usr/local/include
    lib directory: /usr/local/lib
    name: openblas
    openblas configuration: USE_64BITINT= DYNAMIC_ARCH=1 DYNAMIC_OLDER= NO_CBLAS=
      NO_LAPACK= NO_LAPACKE= NO_AFFINITY=1 USE_OPENMP= HASWELL MAX_THREADS=2
    pc file directory: /usr/local/lib/pkgconfig
    version: 0.3.21.dev
  pybind11:
    detection method: config-tool
    include directory: unknown
    name: pybind11
    version: 2.11.1
Compilers:
  c:
    commands: cc
    linker: ld.bfd
    name: gcc
    version: 10.2.1
  c++:
    commands: c++
    linker: ld.bfd
    name: gcc
    version: 10.2.1
  cython:
    commands: cython
    linker: cython
    name: cython
    version: 3.0.8
  fortran:
    commands: gfortran
    linker: ld.bfd
    name: gcc
    version: 10.2.1
  pythran:
    include directory: ../../tmp/pip-build-env-oj5r1eje/overlay/lib/python3.10/site-packages/pythran
    version: 0.15.0
Machine Information:
  build:
    cpu: x86_64
    endian: little
    family: x86_64
    system: linux
  cross-compiled: false
  host:
    cpu: x86_64
    endian: little
    family: x86_64
    system: linux
Python Information:
  path: /opt/python/cp310-cp310/bin/python
  version: '3.10'
```

</details>

### LINKED ISSUE #21010: BUG: Segmentation fault when passing invalid arguments to some functions in the `scipy.linalg.interpolative` module
### Describe your issue.

I was working on Pynguin, an automated unit test generation tool for Python, and the tool found that if we pass invalid arguments to some functions in the `scipy.linalg.interpolative` module, Scipy crashes with a segmentation fault. It's not a particularly serious bug but it might be a good idea to fix it so that it is more user-friendly in the event of a user error.

### Reproducing Code Example

```python
#######################################################################
# Example 1
#######################################################################

from scipy.linalg import interpolative
import numpy

array = numpy.empty(1, dtype=numpy.float64)

# Coming from the function "scipy.linalg._interpolative_backend.iddr_id"
interpolative.interp_decomp(array, 3, False)

#######################################################################
# Example 2
#######################################################################

from scipy.linalg import interpolative
import numpy

array = numpy.empty(1, dtype=numpy.complex128)

# Coming from the function "scipy.linalg._interpolative_backend.idzr_id"
interpolative.interp_decomp(array, 3, False)

#######################################################################
# Example 3
#######################################################################

from scipy.linalg import interpolative
import numpy

array = numpy.array(
    [91, -94, -85, -13, -49, -81, -43, 22, 0, -41],
    dtype=numpy.float64
)

# Coming from the function "scipy.linalg._interpolative_backend.idd_reconint"
interpolative.reconstruct_interp_matrix(array, array)

#######################################################################
# Example 4
#######################################################################

from scipy.linalg import interpolative
import numpy

array = numpy.array(
    [91, -94, -85, -13, -49, -81, -43, 22, 0, -41],
    dtype=numpy.complex128
)

# Coming from the function "scipy.linalg._interpolative_backend.idz_reconint"
interpolative.reconstruct_interp_matrix(array, array)

#######################################################################
# Bonus example 1
#######################################################################

from scipy.linalg import interpolative
import numpy

array = numpy.array([0, -4], dtype=numpy.float64)

# Coming from the function "scipy.linalg._interpolative_backend.idd_reconint"
interpolative.reconstruct_interp_matrix(array, array)

# corrupted size vs. prev_size
# Aborted (core dumped)

#######################################################################
# Bonus example 2
#######################################################################

from scipy.linalg import interpolative
import numpy

array = numpy.array([0, -4], dtype=numpy.float64)

# Coming from the function "scipy.linalg._interpolative_backend.idd_reconint"
interpolative.reconstruct_interp_matrix(array, array).foo

# double free or corruption (!prev)
# Aborted (core dumped)
```

### Error message

```shell
Segmentation fault (core dumped)
```

### SciPy/NumPy/Python version and system information

```shell
1.13.1 2.0.0 sys.version_info(major=3, minor=10, micro=14, releaselevel='final', serial=0)
Build Dependencies:
  blas:
    detection method: pkgconfig
    found: true
    include directory: /usr/local/include
    lib directory: /usr/local/lib
    name: openblas
    openblas configuration: USE_64BITINT=0 DYNAMIC_ARCH=1 DYNAMIC_OLDER= NO_CBLAS=
      NO_LAPACK= NO_LAPACKE= NO_AFFINITY=1 USE_OPENMP= ZEN MAX_THREADS=64
    pc file directory: /usr/local/lib/pkgconfig
    version: 0.3.27
  lapack:
    detection method: pkgconfig
    found: true
    include directory: /usr/local/include
    lib directory: /usr/local/lib
    name: openblas
    openblas configuration: USE_64BITINT=0 DYNAMIC_ARCH=1 DYNAMIC_OLDER= NO_CBLAS=
      NO_LAPACK= NO_LAPACKE= NO_AFFINITY=1 USE_OPENMP= ZEN MAX_THREADS=64
    pc file directory: /usr/local/lib/pkgconfig
    version: 0.3.27
  pybind11:
    detection method: config-tool
    include directory: unknown
    name: pybind11
    version: 2.12.0
Compilers:
  c:
    commands: cc
    linker: ld.bfd
    name: gcc
    version: 10.2.1
  c++:
    commands: c++
    linker: ld.bfd
    name: gcc
    version: 10.2.1
  cython:
    commands: cython
    linker: cython
    name: cython
    version: 3.0.10
  fortran:
    commands: gfortran
    linker: ld.bfd
    name: gcc
    version: 10.2.1
  pythran:
    include directory: ../../tmp/pip-build-env-mnl4e8vy/overlay/lib/python3.10/site-packages/pythran
    version: 0.15.0
Machine Information:
  build:
    cpu: x86_64
    endian: little
    family: x86_64
    system: linux
  cross-compiled: false
  host:
    cpu: x86_64
    endian: little
    family: x86_64
    system: linux
Python Information:
  path: /opt/python/cp310-cp310/bin/python
  version: '3.10'
```

### PR DESCRIPTION
**Pull request by ilayn**
#### Reference issue

Closes #4275
Closes #9249
Closes #12593
Closes #18367
Closes #20452 
Closes #20609
Closes #21010 

Part of #18566

#### What does this implement/fix?
Replaces the infamous id_dist code with its Python equivalent.
The fortran code had a bug with `32bit` systems running `np.complex128` problems. That is now also fixed.

#### Additional information
The performance difference is about 8x slower than the original version at worst. Frankly, this code should never made its way into SciPy in its current form. The method is pretty much rock solid but the code is very haphazard.

There are also additional functions made public which are `seed` and `rand`. The new code is accepting `Generator`s so they are deprecated.

### CODE REVIEW

**Comment by ilayn** [comment]
I'm currently traveling for work so apologies for the piece meal code written mostly on the planes. I'll finish it off quicker when I am at the sea level.

File: scipy/linalg/interpolative.py [hunk 1]
```diff
@@ -424,65 +433,45 @@ def _is_real(A):
...
-        elif state.min() < 0 or state.max() > 1:
-            raise ValueError("values not in range [0,1]")
-        _backend.id_srandi(state)
-    elif seed is None:
-        _backend.id_srandi(np.random.rand(55))
-    else:
-        rnd = np.random.RandomState(seed)
-        _backend.id_srandi(rnd.rand(55))
+    warnings.warn("`scipy.linalg.interpolative.seed` is deprecated and will be "
+                  "removed in SciPy 1.17.0.", DeprecationWarning, stacklevel=2)
```
Comment (ilayn) [inline scipy/linalg/interpolative.py:444]: This is a no-op function now. The stacklevel should be corrected and should be added to the deprecation tracker.

File: scipy/linalg/interpolative.py [hunk 2]
```diff
@@ -424,65 +433,45 @@ def _is_real(A):
...
     Parameters
     ----------
     *shape
         Shape of output array
 
     """
-    # For details, see :func:`_backend.id_srand`, and :func:`_backend.id_srando`.
-    return _backend.id_srand(np.prod(shape)).reshape(shape)
+    warnings.warn("`scipy.linalg.interpolative.rand` is deprecated and will be "
+                  "removed in SciPy 1.17.0.", DeprecationWarning, stacklevel=2)
```
Comment (ilayn) [inline scipy/linalg/interpolative.py:469]: This is an unwanted function now. The stacklevel should be corrected and should be added to the deprecation tracker.

**Comment by ilayn** [comment]
Locally two failures remaining with a slight violation of the tolerances

```shell
=============================================== short test summary info ===============================================
FAILED scipy\linalg\tests\test_interpolative.py::TestInterpolativeDecomposition::test_real_id_fixed_precision[float64-True-True] - AssertionError:
FAILED scipy\linalg\tests\test_interpolative.py::TestInterpolativeDecomposition::test_svd_fixed_precision[float64-True-True] - AssertionError:
============================================ 2 failed, 45 passed in 2.65s =============================================
```

Other than that this is ready for review. I think there is really no point on working on these functions other than legacy reasons but still some housekeeping will be added on top.

**Review by lucascolley — COMMENTED** [review]
`mypy` error just needs line 143 of `mypy.ini` adjusted to the new extension name. The lint failure is from unicode-check, and the smoke-docs failures look like some docstring adjustments are needed.

**Comment by ilayn** [comment]
Well I am happy with that much failure :) Seems like this is going to fly after all.

**Comment by ilayn** [comment]
Apart from the failing array API test, this is sort of ready. I'd appreciate if we don't squash merge it in case there are no comments.

**Comment by ilayn** [comment]
@rkern Just out of curiosity, I have been using the NumPy C API for arrays to send things around to Lapack etc. For that, thanks to Cython, there is a rather convenient option to reach to the underlying data pointers 

```cython

def somefunc(cnp.ndarray arr):
    cdef double* data_ptr = PyArray_DATA(arr)

    # ... some code later 
    nrm = dnrm2(&n, data_ptr, &one)  # compute its norm
    dscal(&n, &five, data_ptr, &one)  # arr * 5.0 inplace

```

in other words, I can grab the internals of the array, possibly modify things and return it back to the Python world. 

In this PR I was trying to use the same mechanism for the random `Generator`s in Cython. Is it possible that if user gives a `rng` instance, I can grab its internals and use the C Random API via `bitgen_t`? My use cases are almost exclusively generating uniform numbers. 

This is going to be needed in a few places soon; especially while sending things to C hence I was wondering if that is possible or not.

**Comment by rkern** [comment]
[Yup!](https://numpy.org/doc/1.26/reference/random/extending.html#cython)

**Comment by ilayn** [comment]
But in those examples we don't have any explicit pass of a `Generator` hence my question. So something along the lines the user passes `rng` created with `rng=np.random.default_rng()`. 

Then do I need to somehow grab the `BitGenerator` of it because that is given in the example? Or are they the same thing?

**Comment by rkern** [comment]
[`Generator.bit_generator`](https://numpy.org/doc/1.26/reference/random/generated/numpy.random.Generator.bit_generator.html) is a public attribute on `Generator`.

**Comment by ilayn** [comment]
Ah OK thank you! I'll play around with it.

**Comment by ilayn** [comment]
This one is ready to go in case there are no further comments.

**Comment by ilayn** [comment]
I've gone through this again in the last two days with random inputs and did not notice any differences between the old fortran code. It turns out `diffsnorm` was broken already in the old code anyways but it seems like nobody is using it hence we did not hear about it. 

So let's see if we get any issues from the nightlies downstream.

**Comment by hoodmane** [comment]
Thanks @ilayn!! I hated compiling `id_dist` soooo much.

> infamous

Indeed.

**Pull request merged**